clients/src/main/java/org/apache/kafka/clients/consumer/internals/CommitRequestManager.java

```diff
@@ -0,0 +1,280 @@
+    private final Optional<CoordinatorRequestManager> coordinatorRequestManager;
+    private final GroupStateManager groupState;
+
+    public CommitRequestManager(
+            final Time time,
+            final LogContext logContext,
+            final SubscriptionState subscriptionState,
+            final ConsumerConfig config,
+            final CoordinatorRequestManager coordinatorRequestManager,
+            final GroupStateManager groupState) {
+        this.log = logContext.logger(getClass());
+        this.stagedCommits = new LinkedList<>();
+        if (config.getBoolean(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG)) {
+            final long autoCommitInterval =
+                    Integer.toUnsignedLong(config.getInt(ConsumerConfig.AUTO_COMMIT_INTERVAL_MS_CONFIG));
+            this.autoCommitState = Optional.of(new AutoCommitState(time, autoCommitInterval));
+        } else {
+            this.autoCommitState = Optional.empty();
+        }
+        this.coordinatorRequestManager = Optional.ofNullable(coordinatorRequestManager);
+        this.groupState = groupState;
+        this.subscriptionState = subscriptionState;
+    }
+
+    // Visible for testing
+    CommitRequestManager(
+            final Time time,
+            final LogContext logContext,
+            final SubscriptionState subscriptionState,
+            final ConsumerConfig config,
+            final CoordinatorRequestManager coordinatorRequestManager,
+            final GroupStateManager groupState,
+            final AutoCommitState autoCommitState) {
+        this.log = logContext.logger(getClass());
+        this.subscriptionState = subscriptionState;
+        this.coordinatorRequestManager = Optional.ofNullable(coordinatorRequestManager);
+        this.groupState = groupState;
+        this.autoCommitState = Optional.ofNullable(autoCommitState);
+        this.stagedCommits = new LinkedList<>();
+    }
+
+    /**
+     * Poll for the commit request if there's any. The function will also try to autocommit, if enabled.
+     *
+     * @param currentTimeMs
+     * @return
+     */
+    @Override
+    public NetworkClientDelegate.PollResult poll(final long currentTimeMs) {
+        if (!coordinatorRequestManager.isPresent()) {
+            return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, new ArrayList<>());
+        }
+
+        maybeAutoCommit(currentTimeMs);
+
+        if (stagedCommits.isEmpty()) {
+            return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, new ArrayList<>());
+        }
+
+        List<NetworkClientDelegate.UnsentRequest> unsentCommitRequests =
+                stagedCommits.stream().map(StagedCommit::toUnsentRequest).collect(Collectors.toList());
+        return new NetworkClientDelegate.PollResult(Long.MAX_VALUE, Collections.unmodifiableList(unsentCommitRequests));
+    }
+
+    public RequestFuture<ClientResponse> add(final Map<TopicPartition, OffsetAndMetadata> offsets,
```

Comment:
Can we use a normal `CompletableFuture` instead of `RequestFuture`. The latter is bound up with `ConsumerNetworkClient` which we are not using in this implementation.

Also, I'm surprised to see this return something as low level as `ClientResponse`. I think we can leave the response handling confined to this class and just expose `CompletableFuture<Void>` or something like that.